In [3]:
import numpy as np
import pandas as pd

### Question 1
What's the version of NumPy that you installed?

In [4]:
np.__version__

'1.21.5'

In [5]:
car_data = pd.read_csv('data/car.csv')
car_data.head()

Unnamed: 0 Make       Model  Year             Engine Fuel Type  Engine HP  \
0           0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1           1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2           2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3           3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4           4  BMW    1 Series  2011  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   
2                Luxury,High-Performance      Compact         Coupe   
3                     Luxury,Performance      Compact         Coupe   
4                                 Luxury      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650  
2           28        20        3916  36350  
3           28        18        3916  29450  
4           28        18        3916  34500

In [6]:
car_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11914 entries, 0 to 11913
Data columns (total 17 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         11914 non-null  int64  
 1   Make               11914 non-null  object 
 2   Model              11914 non-null  object 
 3   Year               11914 non-null  int64  
 4   Engine Fuel Type   11911 non-null  object 
 5   Engine HP          11845 non-null  float64
 6   Engine Cylinders   11884 non-null  float64
 7   Transmission Type  11914 non-null  object 
 8   Driven_Wheels      11914 non-null  object 
 9   Number of Doors    11908 non-null  float64
 10  Market Category    8172 non-null   object 
 11  Vehicle Size       11914 non-null  object 
 12  Vehicle Style      11914 non-null  object 
 13  highway MPG        11914 non-null  int64  
 14  city mpg           11914 non-null  int64  
 15  Popularity         11914 non-null  int64  
 16  MSRP               119

### Question 2
How many records are in the dataset?<br>

Here you need to specify the number of rows.<br>

**16** <br>
**6572**<br>
**11914**<br>
**18990**

In [7]:
print(f'There are {(car_data.shape)[0]} rows')

There are 11914 rows


### Question 3
Who are the most popular car manufacturers (top-3) according to the dataset? <br>

Chevrolet, Volkswagen, Toyota <br>
Chevrolet, Ford, Toyota <br>
Ford, Volkswagen, Toyota <br>
Chevrolet, Ford, Volkswagen

In [8]:
car_by_make = car_data.groupby('Make').agg({'Popularity': 'sum'}).sort_values('Popularity',ascending=False)
car_by_make.head(3)

Popularity
Make                 
Ford          4983817
Chevrolet     1555355
Toyota        1515126

### Question 4 <br>
What's the number of unique Audi car models in the dataset? <br>

3 <br>
16 <br>
26 <br>
34 <br>

In [9]:
Audi_car_data = car_data[car_data['Make'] == 'Audi']
Audi_car_data['Model'].nunique()

34

### Question 5<br>
How many columns in the dataset have missing values?<br>

5 <br>
6<br>
7<br>
8

In [10]:
car_data.isna().sum()

Unnamed: 0              0
Make                    0
Model                   0
Year                    0
Engine Fuel Type        3
Engine HP              69
Engine Cylinders       30
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      3742
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

### Question 6<br>
Find the median value of "Engine Cylinders" column in the dataset.<br>
Next, calculate the most frequent value of the same "Engine Cylinders".<br>
Use the fillna method to fill the missing values in "Engine Cylinders" with the most frequent value from the previous step.<br>
Now, calculate the median value of "Engine Cylinders" once again.
Has it changed?<br>

Hint: refer to existing mode and median functions to complete the task.<br>

Yes<br>
No

In [11]:
print(f'The current median value for Engine Cylinders: {car_data["Engine Cylinders"].median()}')

The current median value for Engine Cylinders: 6.0


In [12]:
print('The most frequent value for Engine Cylinders: ')
print(car_data["Engine Cylinders"].mode())

The most frequent value for Engine Cylinders: 
0    4.0
Name: Engine Cylinders, dtype: float64


In [13]:
car_data["Engine Cylinders"].fillna(4.0, inplace=True)
car_data.isna().sum()

Unnamed: 0              0
Make                    0
Model                   0
Year                    0
Engine Fuel Type        3
Engine HP              69
Engine Cylinders        0
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      3742
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

In [14]:
print(f'The updated median value for Engine Cylinders: {car_data["Engine Cylinders"].median()}')

The updated median value for Engine Cylinders: 6.0


### Question 7<br>
Select all the "Lotus" cars from the dataset.<br>
Select only columns "Engine HP", "Engine Cylinders".<br>
Now drop all duplicated rows using drop_duplicates method (you should get a dataframe with 9 rows).<br>
Get the underlying NumPy array. Let's call it X.<br>
Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.<br>
Invert XTX.<br>
Create an array y with values [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800].<br>
Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.<br>
What's the value of the first element of w?<br>
Note: You just implemented linear regression. We'll talk about it in the next lesson.<br>

-0.0723<br>
4.5949<br>
31.6537<br>
63.5643

In [18]:
lotus_cars = car_data[car_data['Make'] == 'Lotus']
X = lotus_cars[["Engine HP", "Engine Cylinders"]].drop_duplicates().to_numpy()
XTX = np.dot(X.T, X) 
XTX

array([[7.31684e+05, 1.34100e+04],
       [1.34100e+04, 2.52000e+02]])

In [19]:
Y = np.array([1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800])
Y

array([1100,  800,  750,  850, 1300, 1000, 1000, 1300,  800])

In [21]:
from numpy.linalg import inv
W = np.dot(np.dot(inv(XTX) , X.T) * Y)
W

TypeError: dot() missing 1 required positional argument: 'b'